### PR TITLE
Add Segment SaaS Connector + Authentication per Request  [#251] 

### LINKED ISSUE #251: [SaaS Connector] Segment
Create an integration with segment to process privacy requets

Mohan has translated the legacy yaml, so please reference that when starting this work.

See [requirements](https://ethyca.atlassian.net/wiki/spaces/OSS/pages/2211709180/Segment+Connector+Specification)

### PR DESCRIPTION
**Pull request by pattisdr**
# Purpose

- Adds example support for connecting to third-party infrastructure: **Segment** + adds the ability to configure custom authentication for individual requests. 
- Demonstrates end-to-end access request for Segment.

Segment masking will be demonstrated in #354 

# Changes
- Adds example Segment saas config (for annotating how to access Segment endpoints) and Segment dataset (for annotating what fields we have)
-   - Adds GDPR delete request under `data_protection_request` (needs #325 to work, but this will be tested in follow-up #354)
- Currently, we can only set a ClientConfig for the entire SaaSConnector, which contains things like the domain and the authentication strategy to connect to various third party endpoints.  However, with Segment, different endpoints have different domains and authentication strategies (bearer and basic)**.  Allow setting authentication at the individual request level.** We will check if authentication exists at the request level first, otherwise, we'll fall back to the overall ClientConfig set up for the SaaSConnector.
- Adds end to end test for **access request** only

# Checklist

- [x] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md))
- [x] Good unit test/integration test coverage
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [x] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket
Uses #328 
Fixes #251

### CODE REVIEW

File: data/saas/config/segment_config.yml [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+              username:
+                connector_param: access_secret
+  - name: traits
+    requests:
+      read:
+        method: GET
+        path: /v1/spaces/<namespace_id>/collections/users/profiles/<segment_id>/traits
+        query_params:
+          - name: limit
+            value: 17
```
Comment (pattisdr) [inline data/saas/config/segment_config.yml:84]: We can get all the attributes in one request if we specify limit=17 here.

File: Makefile [hunk 2]
```diff
@@ -124,7 +124,7 @@ pytest-integration-external: compose-build
 
 pytest-saas: compose-build
 	@echo "Running integration tests for SaaS connectors"
-	@docker-compose run -e MAILCHIMP_DOMAIN -e MAILCHIMP_USERNAME -e MAILCHIMP_API_KEY -e MAILCHIMP_IDENTITY_EMAIL -e SENTRY_HOST -e SENTRY_ACCESS_TOKEN -e SENTRY_IDENTITY_EMAIL  $(IMAGE_NAME) \
+	@docker-compose run -e MAILCHIMP_DOMAIN -e MAILCHIMP_USERNAME -e MAILCHIMP_API_KEY -e MAILCHIMP_IDENTITY_EMAIL -e SENTRY_HOST -e SENTRY_ACCESS_TOKEN -e SENTRY_IDENTITY_EMAIL -e SEGMENT_DOMAIN -e SEGMENT_PERSONAS_DOMAIN -e SEGMENT_WORKSPACE -e SEGMENT_ACCESS_TOKEN -e SEGMENT_NAMESPACE_ID -e SEGMENT_ACCESS_SECRET -e SEGMENT_IDENTITY_EMAIL  $(IMAGE_NAME) \
```
Comment (pattisdr) [inline Makefile:127]: Will consolidate with Adrian's cleanup of these after his is merged -

File: data/saas/config/segment_config.yml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    body: '{\"regulation_type\": \"Suppress_With_Delete\", \"attributes\": {\"name\": \"userId\",\"values\": [\"<user_id>\",]}}'
+    client_config:
+      protocol: https
+      host:
+        connector_param: config_domain
+      authentication:
+        strategy: bearer_authentication
+        configuration:
+          username:
+            connector_param: access_token
```
Comment (pattisdr) [inline data/saas/config/segment_config.yml:154]: This has several fixes in the erasure PR, I'd ignore for now

File: data/saas/config/segment_config.yml [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            identity: email
+        client_config:
+          protocol: https
+          host:
+            connector_param: personas_domain
+          authentication:
+            strategy: basic_authentication
+            configuration:
+              username:
+                connector_param: access_secret
```
[the remark this answers is not in the document]
Comment (galvana) [reply data/saas/config/segment_config.yml:48]: Thanks for adding this! I like that you added the entire `client_config` in here, I was thinking of adding a custom `authentication` strategy but your approach is more flexible.

File: docs/fidesops/docs/postman/Fidesops.postman_collection.json [hunk 5]
```diff
@@ -2535,6 +2535,229 @@
 						}
 					]
 				},
+				{
+					"name": "Segment",
```
Comment (galvana) [inline docs/fidesops/docs/postman/Fidesops.postman_collection.json:2539]: This is a helpful addition, do you think we should add a new collection for each integration we add?

File: data/saas/dataset/segment_dataset.yml [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    name: Segment Dataset
+    description: A sample dataset representing the Segment connector for Fidesops
+    collections:
+      - name: segment_user
+        fields:
+          - name: segment_id
+            data_categories: [user.derived.identifiable.unique_id]
+            fidesops_meta:
+              primary_key: True
+      - name: track_events
```
Comment (galvana) [inline data/saas/dataset/segment_dataset.yml:12]: Do you have any opinions around standardizing the collection names? My preference would be to use singular names for the collections.

**Comment by seanpreston** [comment]
For the CodeQL alert, should we at least be including a check to a config var. similar to `LOG_PII` for connection secrets if we really have to log it?

File: docs/fidesops/docs/postman/Fidesops.postman_collection.json (hunk 5 above)
Comment (pattisdr) [reply docs/fidesops/docs/postman/Fidesops.postman_collection.json:2539]: I mean I don't want to go overboard, but this was helpful for myself when I was testing.  I also duplicated the collection and edited for Stripe when I was testing your code. I think it's useful in the beginning.

File: data/saas/dataset/segment_dataset.yml (hunk 6 above)
Comment (pattisdr) [reply data/saas/dataset/segment_dataset.yml:12]: I think these were piped over from the historical segment yaml - I guess I don't understand why we need a standard internally here?  To me it's less confusing to match the third party source we are connecting to, then arbitrarily making them singular.

File: data/saas/config/segment_config.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (pattisdr) [reply data/saas/config/segment_config.yml:48]: the driver was that I needed a new host too for these endpoints!

File: src/fidesops/service/pagination/pagination_strategy_link.py [hunk 7]
```diff
@@ -50,7 +50,7 @@ def get_next_request(
         elif self.source == LinkSource.body.value:
             next_link = pydash.get(response.json(), self.path)
 
-        if next_link is None:
+        if not next_link:
```
Comment (galvana) [inline src/fidesops/service/pagination/pagination_strategy_link.py:53]: Was the scenario you ran into here accounted for in the tests? If not, can we add a test?

Comment (pattisdr) [reply src/fidesops/service/pagination/pagination_strategy_link.py:53]: It's accounted for end-to-end in the segment tests! For segment, next links are empty strings if they don't exist, so this was causing a bad url to be built and the traversal to fail.

Comment (pattisdr) [reply src/fidesops/service/pagination/pagination_strategy_link.py:53]: test added!

File: data/saas/dataset/segment_dataset.yml (hunk 6 above)
Comment (pattisdr) [reply data/saas/dataset/segment_dataset.yml:12]: Discussed, leaving as-is for now, we may have a broader discussion about this Friday.

File: tests/service/pagination/test_pagination_strategy_link.py [hunk 8]
```diff
@@ -101,6 +116,21 @@ def test_link_in_body_missing(response_with_header_link):
     assert next_request is None
 
 
+def test_link_in_body_empty_string(response_with_empty_string_link):
```
Comment (galvana) [inline tests/service/pagination/test_pagination_strategy_link.py:119]: Thank you 🙏

**Review by galvana — APPROVED** [review]
Thanks for answering my questions and adding the extra test. One more connector down but most importantly glad you added the client config override, that's going to be very useful for other SaaS connectors!

**Pull request merged**